File: app/src/main/java/com/vo1d/schedulemanager/v2/data/subject/SubjectsDao.java
package com.vo1d.schedulemanager.v2.data.subject;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;
import androidx.room.Transaction;

import com.vo1d.schedulemanager.v2.data.BaseDao;

import java.util.List;

@Dao
public abstract class SubjectsDao implements BaseDao<Subject> {

    @Transaction
    @Query("DELETE FROM subject_table")
    abstract void deleteAll();

    @Transaction
    @Query("SELECT * FROM subject_table ORDER BY title ASC")
    abstract LiveData<List<Subject>> getAllSubjects();

    @Transaction
    @Query("SELECT * FROM subject_table WHERE title LIKE '%' || :filter || '%' ORDER BY title ASC")
    abstract List<Subject> getFilteredSubjects(String filter);

    @Query("SELECT * FROM subject_table WHERE id=:id")
    abstract Subject findSubjectById(int id);
}

File: app/src/main/kotlin/ru/vo1d/timetablemanager/data/entities/subjects/SubjectsDao.kt
package ru.vo1d.timetablemanager.data.entities.subjects

import androidx.room.Dao
import androidx.room.Query
import kotlinx.coroutines.flow.Flow
import ru.vo1d.timetablemanager.data.BaseDao
import ru.vo1d.timetablemanager.data.entities.sessions.SessionType

@Dao
interface SubjectsDao : BaseDao<Int, Subject> {
    @get:Query("SELECT * FROM subjects ORDER BY title ASC")
    val all: Flow<List<Subject>>


    @Query(
        """SELECT * FROM subjects
        WHERE title LIKE '%' || :filter || '%'
        ORDER BY title ASC"""
    )
    fun find(filter: String): Flow<List<Subject>>

    @Query("SELECT types FROM subjects WHERE id = :id")
    fun findTypesFor(id: Int): Flow<List<SessionType>>


    @Query("SELECT * FROM subjects WHERE id = :id")
    suspend fun findById(id: Int): Subject?

    @Query("DELETE FROM subjects")
    suspend fun deleteAll()
}